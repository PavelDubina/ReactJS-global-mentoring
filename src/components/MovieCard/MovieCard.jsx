import React from 'react'
import PropTypes from 'prop-types'
import { ImageWithFallback } from '../ImageWithFallback/ImageWithFallback'
import styles from './MovieCard.scss'

export const MovieCard = ({ title, genres, date, poster }) => (
  <div className={styles.container}>
    <div className={styles.poster}>
      <ImageWithFallback src={poster} />
      {/* <img width="9" height="16" src={poster} alt="poster" /> */}
      <div className={styles.results_filter}>
        <div />
        <div />
        <div />
      </div>
    </div>
    <div className={styles.movie__data}>
      <div className={styles.movie__data_info}>
        <h2 className={styles.movie__data_info_title}>{title}</h2>
        <p className={styles.movie__data_info_genres}>{genres}</p>
      </div>
      <div className={styles.movie__data_date}>
        <span>{date}</span>
      </div>
    </div>
  </div>
)

MovieCard.propTypes = {
  title: PropTypes.string.isRequired,
  genres: PropTypes.string.isRequired,
  date: PropTypes.number.isRequired,
  poster: PropTypes.string.isRequired,
}
