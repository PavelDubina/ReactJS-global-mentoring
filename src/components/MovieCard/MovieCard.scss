.container {
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
  background-color: #fff;
}

.menu {
  display: none;

  &_opened {
    display: block;
  }
}

.poster {
  position: relative;
  padding-top: 148%;
  overflow: hidden;
  cursor: pointer;

  & img {
    position: absolute;
    top: 0;
    min-width: 100%;
    height: auto;
    object-fit: cover;
    color: #fff;
    text-align: center;
  }

  &:hover .menu {
    display: block;
  }
}

.movie__data {
  padding: 8px 8px 0 8px;
  display: flex;
  justify-content: space-between;
  background-color: #232323;

  &_info {
    display: flex;
    flex-direction: column;
    overflow-x: hidden;

    &_title {
      margin: 0 0 4px 0;
      font-size: 18px;
      overflow-x: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      color: #ffffff7a;
    }

    &_genres {
      font-size: 14px;
      overflow-x: hidden;
      text-overflow: ellipsis;
      text-transform: capitalize;
      white-space: nowrap;
      color: #555;
    }
  }

  &_date {
    font-size: 14px;
    margin: 5px 0 0 0;
    color: #ffffff7a;

    & span {
      padding: 0 8px;
      border: 1px solid #555;
      border-radius: 5px;
    }
  }
}
