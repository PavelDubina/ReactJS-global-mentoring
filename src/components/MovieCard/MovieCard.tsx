import React, { useState } from 'react'
import className from 'classnames'
import { CardMenu } from './CardMenu/CardMenu'
import { getReleaseYear, getCorrectGenres } from '../../utils/helpers'
import { ImageWithFallback } from '../ImageWithFallback/ImageWithFallback'
import styles from '../../../styles/MovieCard.module.scss'

type MovieCardProps = {
  title: string
  genres: string[]
  date: string
  poster?: string
  id: number
  getMovie: (id: number, status: string, isOpen: boolean) => void
  handleMovieDetails: (id: number) => void
}

export const MovieCard: React.FC<MovieCardProps> = ({
  title,
  genres,
  date,
  poster,
  id,
  getMovie,
  handleMovieDetails,
}) => {
  const [isOpen, setIsOpen] = useState(false)
  const menuStyle = className(styles.menu, { [styles.menu_opened]: isOpen })
  const toggleMenu = () => setIsOpen((state) => !state)
  const onOpenModal = (status: string) => {
    getMovie(id, status, true)
    setIsOpen((state) => !state)
  }
  return (
    <>
      <div className={styles.container}>
        <div className={styles.poster}>
          <ImageWithFallback onClick={() => handleMovieDetails(id)} src={poster} />
          <div data-testid="menu" className={menuStyle}>
            <CardMenu isOpen={isOpen} onOpenModal={onOpenModal} toggleMenu={toggleMenu} />
          </div>
        </div>
        <div className={styles.movie__data}>
          <div className={styles.movie__data_info}>
            <h2 className={styles.movie__data_info_title}>{title}</h2>
            <p className={styles.movie__data_info_genres}>{getCorrectGenres(genres)}</p>
          </div>
          <div className={styles.movie__data_date}>
            <span>{getReleaseYear(date)}</span>
          </div>
        </div>
      </div>
    </>
  )
}
