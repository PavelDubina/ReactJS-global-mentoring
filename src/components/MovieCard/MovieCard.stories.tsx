import React from 'react'
import { Story, Meta } from '@storybook/react'
import { MovieCard, MovieCardProps } from './MovieCard'

export default {
  title: 'MyStorybook/MovieCard',
  component: MovieCard,
} as Meta

const Template: Story<MovieCardProps> = (args) => (
  <div style={{ width: '260px', height: '340px' }}>
    <MovieCard {...args} />
  </div>
)

export const Default = Template.bind({})
Default.args = {
  title: 'Avengers',
  genres: ['comedy', 'action', 'adventure'],
  date: '2013',
  poster: 'https://image.tmdb.org/t/p/w500/9QA8rjbMqPcJXdHWEZE4Nxvktgx.jpg',
  id: 1,
  getMovie: () => null,
  handleMovieDetails: () => null,
}
