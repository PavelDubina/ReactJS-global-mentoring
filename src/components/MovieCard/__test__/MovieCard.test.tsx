import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import { MovieCard } from '../MovieCard'

describe('MovieCard', () => {
  const getComponent = ({
    title = 'test title',
    genres = ['comedy'],
    date = '12/12/2020',
    poster = 'test src',
    id = 122,
    getMovie = jest.fn(),
    handleMovieDetails = jest.fn(),
  }) =>
    render(
      <MovieCard
        title={title}
        genres={genres}
        date={date}
        poster={poster}
        id={id}
        getMovie={getMovie}
        handleMovieDetails={handleMovieDetails}
      />,
    )
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
  it('Should have class menu_opened', () => {
    const { getByTestId, getByText } = getComponent({})
    fireEvent.click(getByTestId('card-menu'))
    expect(getByTestId('menu').classList.contains('menu_opened')).toBeTruthy()
    fireEvent.click(getByText('menu.edit'))
    expect(getByTestId('menu').classList.contains('menu_opened')).toBeFalsy()
  })
  it('Sould to be called handleMovieDetails', () => {
    const handleMovieDetails = jest.fn()
    const { getByAltText } = getComponent({ handleMovieDetails })
    fireEvent.click(getByAltText('poster'))
    expect(handleMovieDetails).toBeCalled()
  })
})
