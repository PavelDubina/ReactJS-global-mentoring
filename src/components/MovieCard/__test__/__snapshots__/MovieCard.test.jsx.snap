// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="poster"
    >
      <img
        alt="poster"
        height="16"
        src="test src"
        width="9"
      />
      <div
        class="menu"
        data-testid="menu"
      >
        <div
          class="menu"
          data-testid="card-menu"
        >
          <div />
          <div />
          <div />
        </div>
      </div>
    </div>
    <div
      class="movie__data"
    >
      <div
        class="movie__data_info"
      >
        <h2
          class="movie__data_info_title"
        >
          test title
        </h2>
        <p
          class="movie__data_info_genres"
        >
          comedy
        </p>
      </div>
      <div
        class="movie__data_date"
      >
        <span>
          2020
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
